test/RelayServer.test.js

```diff
@@ -667,6 +670,22 @@ contract('RelayServer', function (accounts) {
       const promises = [relayTransaction(options), relayTransaction(options2)]
       await Promise.all(promises)
     })
+    it('should not deadlock if server returned error while locked', async function () {
+      try {
+        relayServer.keyManager.signTransactionOrig = relayServer.keyManager.signTransaction
+        relayServer.keyManager.signTransaction = function () {
+          throw new Error('no tx for you')
+        }
+        try {
+          await relayTransaction(options)
+        } catch (e) {
+          assert.equal(e.message, 'no tx for you', e.message)
+          assert.isFalse(relayServer.nonceMutex.isLocked(), 'nonce mutex not released after exception')
```

Comment:
but this is quite a white-box test... it would be more articulate if you made a second 'relayTransaction' request and assert your computer did not catch fire